### ISSUE TITLE
v1.15 Operator deletes and creates EndpointSlices on startup

### ISSUE THREAD

**Issue by dctrwatson** [cilium/cilium#31564]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

First noticed that upgrading from v1.14 to v1.15 this happens. But now have had it happen even when v1.15 operator is just restarted.

If there are enough endpoints in the cluster and network policies exist. Pods lose connectivity for minutes at a time due to the delay between Delete and Create events.

### Cilium Version

1.15.2 7cf57829 2024-03-13T15:34:43+02:00 go version go1.21.8 linux/amd64

### Kernel Version

5.15.0-1055-aws #60~20.04.1-Ubuntu

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"26+", GitVersion:"v1.26.13-eks-508b6b3", GitCommit:"26dbbc7674e3d15b579cec35cdd6aede55c1d426", GitTreeState:"clean", BuildDate:"2024-01-29T20:58:43Z", GoVersion:"go1.20.13", Compiler:"gc", Platform:"linux/amd64"}

### Regression

v1.14.x

### Anything else?

https://github.com/cilium/cilium/pull/26836 seems relevant

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by chaunceyjiang** [comment]
Have you tried adjusting the `ces-*` related parameters?

**Comment by dctrwatson** [comment]
> Have you tried adjusting the `ces-*` related parameters?

No. Everything is default for those.

**Comment by marseel** [comment]
@dctrwatson could you provide logs from cilium-operator?

This will help us to understand what is going on.

/cc @alan-kut @dlapcevic

**Comment by aojea** [comment]
> @dctrwatson could you provide logs from cilium-operator?

and also one of the agents logs

> ods lose connectivity for minutes at a time due to the delay between Delete and Create events.

how do you measure those delays?

**Comment by dctrwatson** [comment]
> > @dctrwatson could you provide logs from cilium-operator?
> 
> and also one of the agents logs

Unfortunately, there are no logs at the warn level.
We've paused all cilium changes until we can reproduce in a test environment. I'll make sure to enable info logging there.

> how do you measure those delays?

We can see the delay from connection errors reported by services. They line up with policy deny in hubble logs.

Overall, we see drops until creation events return to normal per agent metrics.

For example, we have connections errors for the 10m between when deletes start and until creates stop:
![Screenshot 2024-04-01 at 4 04 06 PM](https://github.com/cilium/cilium/assets/395054/c3099b15-38c1-45d3-ad00-7c70477aa032)

**Comment by dctrwatson** [comment]
> @dctrwatson could you provide logs from cilium-operator?

Confirmed it's deleting all CES every start-up.

The thing that stands out to me is this log message:
https://github.com/cilium/cilium/blob/9b3f9a8c0d71bf70bc0fd3fbc256bacc67fe6c32/operator/pkg/ciliumendpointslice/manager.go#L200-L206

I know their identities are not changing. I also confirmed that the identity is the same by taking a snapshot of CES before and after restarts too.

So it looks like all existing CES are updated to have no CEPs so they're deleted.

We use an opt-in list of labels for identities if that somehow changes the way the operator determines identities.

> and also one of the agents logs

As those CES deletes and subsequent adds are reconciled, the agents are processing them.

Approx `2024-04-12 22:38:00.152`
```
level=debug msg="CES deleted, calling endpointDeleted" CEPName=kube-system/snapshot-controller-5bb67dc979-z2mxq CESName=ces-shgdb76g8-mmn7j subsys=k8s-watcher
level=info msg="CEP deleted, calling endpointDeleted" CEPName=kube-system/snapshot-controller-5bb67dc979-z2mxq CESName=ces-shgdb76g8-mmn7j subsys=k8s-watcher
level=debug msg="Deleting IP from ipcache layer" ipAddr=10.0.168.15 subsys=ipcache
level=debug msg="Daemon notified of IP-Identity cache state change" identity="{80827 custom-resource [] false false}" ipAddr="{10.0.168.15 ffffffff}" modification=Delete subsys=datapath-ipcache
```

Approx `2024-04-12 22:38:04.101`:
```
level=debug msg="CES added, calling CoreEndpointUpdate" CEPName=kube-system/snapshot-controller-5bb67dc979-z2mxq CESName=ces-sn8z7hb9d-r9nt6 subsys=k8s-watcher
level=debug msg="Upserting IP into ipcache layer" identity="{80827 custom-resource [] false false}" ipAddr=10.0.168.15 k8sNamespace=kube-system k8sPodName=snapshot-controller-5bb67dc979-z2mxq key=0 namedPorts="map[]" subsys=ipcache
level=debug msg="Daemon notified of IP-Identity cache state change" identity="{80827 custom-resource [] false false}" ipAddr="{10.0.168.15 ffffffff}" modification=Upsert subsys=datapath-ipcache
```

Even in this small cluster, that's at least 4s of dropped traffic.

**Comment by dlapcevic** [comment]
I don't know what the root cause is here, but you could try changing the CES batching mode.
The default one is still the Identity batching mode (group by Identities), while there is also FCFS (First Come First Serve) mode that we're going to make the default. There is a flag `ces-slice-mode` that controls the batching mode. You can set it to any value other than `cesSliceModeIdentity`, and the FCFS mode will be active.

Also, I have a question. How come the Identity is 80827, which is above 65k? Is it using remote identities?

**Comment by dctrwatson** [comment]
> There is a flag `ces-slice-mode` that controls the batching mode. You can set it to any value other than `cesSliceModeIdentity`, and the FCFS mode will be active.

Let me change that and see if it makes CES stable across restarts.
Is this flag available in v1.14?

> Also, I have a question. How come the Identity is 80827, which is above 65k? Is it using remote identities?

This is a test cluster in a single member ClusterMesh. We don't have remote workloads either.
We don't GC identities for 24h to reduce the CPU usage of the agent, so there may be just that many identities per day.

**Comment by dctrwatson** [comment]
> There is a flag `ces-slice-mode` that controls the batching mode. You can set it to any value other than `cesSliceModeIdentity`, and the FCFS mode will be active.

This makes CES stable whenever the v1.15 operator restarts and even when upgrading/downgrading v1.14 to v1.15.

So I think that solves this issue, thank you.
